### DISCUSSION TITLE
Foundry Fundamentals - Section 4: Lesson 39 - One Big Test

### DISCUSSION THREAD

**Discussion by LordAdhaar** [Cyfrin/foundry-full-course-cu#2798] [Q&A]
function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public {
         
        // line 247 
        uint256 additionalEntrances = 3;
        uint256 startingIndex = 1; // We have starting index be 1 so we can start with address(1) and not address(0)

        for (uint256 i = startingIndex; i < startingIndex + additionalEntrances; i++) {
            address player = address(uint160(i));
            hoax(player, 1 ether); // deal 1 eth to the player
            raffle.enterRaffle{value: raffleEntranceFee}();
        }
        
        // line 273
        uint256 prize = raffleEntranceFee * (additionalEntrances + 1);
        }
        
Why is prize calculated using four players (3+1) but the for loop runs for only 3 iterations (1<4; 1,2,3)
I believe my understanding about the code is wrong.

**Comment by EngrPips** [comment]
Hello @LordAdhaar, I am unsure what is going on in your codebase but below is my test 

```js
 function testFulfillRandomWordsPicksAWinnerResetAndSendMoney() public skipFork {
        uint256 totalNumberOfParticipant = 6;
        uint160 startingNumber = 1;

        for(startingNumber; startingNumber < totalNumberOfParticipant; startingNumber++) {
            address cureentPlayer = address(startingNumber);
            hoax(cureentPlayer, startingBalance);
            lottery.enter{value: _entryFee}();
        }

        uint256 prize = _entryFee * (totalNumberOfParticipant - 1);

        vm.warp(block.timestamp + _interval + 1);
        vm.roll(block.number + 1);

        vm.recordLogs();
        lottery.performUpkeep("0x0");
        Vm.Log[] memory entries = vm.getRecordedLogs();
        bytes32 requestId = entries[1].topics[1];

        uint256 initialLastTimeStamp = lottery.getLastTimeStamp();

        VRFCoordinatorV2Mock(_vrfCoordinatorContractAddress).fulfillRandomWords(uint256(requestId), address(lottery));

        // CONTRACT VALUES
        address winner = lottery.getRecentWinner();
        uint256 lengthOfParticipantsArray = lottery.getLengthOfParticipantsArray();
        uint256 lastTimeStamp = lottery.getLastTimeStamp();
        uint256 lotteryState = lottery.getLotteryState();
        uint256 balanceOfTheWinner = winner.balance;

        // EXPECTED VALUES
        uint256 expectedlengthOfParticipantsArray = 0;
        uint256 expectedlotteryState = 0;

        assert(winner != address(0));
        assertEq(lengthOfParticipantsArray, expectedlengthOfParticipantsArray);
        assert(lastTimeStamp > initialLastTimeStamp);
        assertEq(lotteryState, expectedlotteryState);
        assert(balanceOfTheWinner == (startingBalance + prize) - _entryFee);
    }
```

**Comment by LordAdhaar** [comment]
I have modified the code in my own codebase so it works but in the official github repo of cyfrin updraft 

` function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEntered skipFork {
        address expectedWinner = address(1);

        // Arrange
        uint256 additionalEntrances = 3;
        uint256 startingIndex = 1; // We have starting index be 1 so we can start with address(1) and not address(0)

        for (uint256 i = startingIndex; i < startingIndex + additionalEntrances; i++) {
            address player = address(uint160(i));
            hoax(player, 1 ether); // deal 1 eth to the player
            raffle.enterRaffle{value: raffleEntranceFee}();
        }

        uint256 startingTimeStamp = raffle.getLastTimeStamp();
        uint256 startingBalance = expectedWinner.balance;

        // Act
        vm.recordLogs();
        raffle.performUpkeep(""); // emits requestId
        Vm.Log[] memory entries = vm.getRecordedLogs();
        console2.logBytes32(entries[1].topics[1]);
        bytes32 requestId = entries[1].topics[1]; // get the requestId from the logs

        VRFCoordinatorV2_5Mock(vrfCoordinatorV2_5).fulfillRandomWords(uint256(requestId), address(raffle));

        // Assert
        address recentWinner = raffle.getRecentWinner();
        Raffle.RaffleState raffleState = raffle.getRaffleState();
        uint256 winnerBalance = recentWinner.balance;
        uint256 endingTimeStamp = raffle.getLastTimeStamp();
        uint256 prize = raffleEntranceFee * (additionalEntrances + 1);

        assert(recentWinner == expectedWinner);
        assert(uint256(raffleState) == 0);
        assert(winnerBalance == startingBalance + prize);
        assert(endingTimeStamp > startingTimeStamp);
    }`

In this code I was wondering why are calculating prize for 4 when the for loop is running for 3 iterations

**Reply by EngrPips** [reply]
Because of the `raffleEntered` modifier, someone entered in the modifier at first plus the three that entered in the main test function which is a total of 4 participants sir.

**Comment by LordAdhaar** [comment]
Makes sense, guess I missed that part 
Thank you once again ser.

**Reply by EngrPips** [reply]
You are welcome, Keep crushing it.
